src/bluetooth-fw/nimble/nimble_store.c

```diff
@@ -0,0 +1,268 @@
+  struct ble_store_key_sec *key_sec = (struct ble_store_key_sec *)data;
+
+  return ble_addr_cmp(&s->value_sec.peer_addr, &key_sec->peer_addr) == 0;
+}
+
+static BleStoreValue *prv_nimble_store_find_sec(const struct ble_store_key_sec *key_sec) {
+  if (!ble_addr_cmp(&key_sec->peer_addr, BLE_ADDR_ANY)) {
+    return (BleStoreValue *)list_get_at((ListNode *)s_value_secs, key_sec->idx);
+  } else if (key_sec->idx == 0) {
+    return (BleStoreValue *)list_find((ListNode *)s_value_secs, prv_nimble_store_find_sec_cb,
+                                      (void *)&key_sec->peer_addr);
+  }
+
+  return NULL;
+}
+
+static int prv_nimble_store_read_our_sec(const struct ble_store_key_sec *key_sec,
+                                         struct ble_store_value_sec *value_sec) {
+  int ret = 0;
+  BleStoreValue *s;
+
+  bt_lock();
+
+  s = prv_nimble_store_find_sec(key_sec);
+  if (s == NULL) {
+    ret = BLE_HS_ENOENT;
+    goto unlock;
+  }
+
+  memset(value_sec, 0, sizeof(*value_sec));
+
+  value_sec->peer_addr = key_sec->peer_addr;
+  value_sec->key_size = KEY_SIZE;
+
+  if (s->value_sec.ltk_present) {
+    value_sec->ediv = s->value_sec.ediv;
+    value_sec->rand_num = s->value_sec.rand_num;
+    value_sec->ltk_present = true;
+    memcpy(value_sec->ltk, s->value_sec.ltk, KEY_SIZE);
+  }
+
+unlock:
+  bt_unlock();
+
+  return ret;
+}
+
+static int prv_nimble_store_read_peer_sec(const struct ble_store_key_sec *key_sec,
+                                          struct ble_store_value_sec *value_sec) {
+  int ret = 0;
+  BleStoreValue *s;
+
+  bt_lock();
+
+  s = prv_nimble_store_find_sec(key_sec);
+  if (s == NULL) {
+    ret = BLE_HS_ENOENT;
+    goto unlock;
+  }
+
+  *value_sec = s->value_sec;
+
+unlock:
+  bt_unlock();
+
+  return ret;
+}
+
+static int prv_nimble_store_write_peer_sec(const struct ble_store_value_sec *value_sec) {
+  BleStoreValue *s;
+  struct ble_store_key_sec key_sec;
+  BleBonding bonding;
+  BTDeviceAddress addr;
+
+  if (value_sec->key_size != KEY_SIZE) {
+    PBL_LOG_D(LOG_DOMAIN_BT, LOG_LEVEL_ERROR, "Unsupported key size %d", value_sec->key_size);
+    return BLE_HS_ENOTSUP;
+  }
+
+  ble_store_key_from_value_sec(&key_sec, value_sec);
+
+  bt_lock();
+
+  s = prv_nimble_store_find_sec(&key_sec);
+  if (s == NULL) {
+    s = kernel_zalloc_check(sizeof(BleStoreValue));
+    if (s_value_secs == NULL) {
+      s_value_secs = s;
+    } else {
+      list_append((ListNode *)s_value_secs, (ListNode *)s);
+    }
+  }
+
+  s->value_sec = *value_sec;
+
+  bt_unlock();
+
+  // inform about new IRK
+  if (value_sec->irk_present) {
+    BleIRKChange irk_change_event;
+
+    irk_change_event.irk_valid = true;
+    memcpy(irk_change_event.irk.data, value_sec->irk, KEY_SIZE);
+
+    nimble_addr_to_pebble_device(&value_sec->peer_addr, &irk_change_event.device);
+
+    bt_driver_handle_le_connection_handle_update_irk(&irk_change_event);
+  }
+
+  // persist bonding
+  memset(&bonding, 0, sizeof(bonding));
+
+  bonding.is_gateway = true;
+
```

Comment:
assert if other things are present that we do not know how to store (CSRK)?

Reply:
(or, at least, have a comment indicating that this would be the place to put it if someone tried to turn this on in NimBLE)

Reply:
added extra checks that will return not supported and log an error